### ISSUE TITLE
RuntimeError: reentrant call inside <_io.BufferedWriter name='<stdout>'>

### ISSUE THREAD

**Issue by jquast** [rolfmorel/macht#1]
If I resize very quickly in rapid succession, (with osx iTerm2, holding down <cloverleaf>+-, performing dozens of resize signaals/second) I get this very obscure error.

This may very well be something to fix in blessed. just need to understand what this is all about.

```
$ ./macht.py -b 3
Traceback (most recent call last):
  File "./macht.py", line 6, in <module>
    exit(main())
  File "/Users/dingo1/Code/macht/macht/term/main.py", line 109, in main
    key = term.inkey()
  File "/Users/dingo1/.virtualenvs/macht/lib/python3.3/site-packages/blessed/terminal.py", line 662, in inkey
    while not ks and self.kbhit(_timeleft(stime, timeout)):
  File "/Users/dingo1/.virtualenvs/macht/lib/python3.3/site-packages/blessed/terminal.py", line 533, in kbhit
    check_r, check_w, check_x, timeout)
  File "/Users/dingo1/Code/macht/macht/term/main.py", line 78, in term_resize
    grid.draw()
  File "/Users/dingo1/Code/macht/macht/term/grid.py", line 33, in draw
    print(style(self.vert_div))
  File "/Users/dingo1/Code/macht/macht/term/main.py", line 48, in term_resize
    print(term.clear())
RuntimeError: reentrant call inside <_io.BufferedWriter name='<stdout>'>
```

**Comment by jquast** [comment]
this resolves it by using a 'dirty' flag. not ideal but it makes it not re-entrant. something to consider anyway, not a proper pull request.

term_resize() currently is blocking when it writes to stdout(). There might be something proper around signal.SIG_IGN, but i'm afraid that'd mean the most recent SIGWINCH is ignored in this case.

```
--- a/macht/term/main.py
+++ b/macht/term/main.py
@@ -80,6 +80,8 @@ def term_resize(term, grids, signum=None, frame=None):


 def main(args=None):
+    global do_refresh
+    do_refresh = True
     opts = parser.parse_args(args or sys.argv[1:])

     term = blessed.Terminal()
@@ -96,17 +98,22 @@ def main(args=None):

         grids.append(grid)

-    signal.signal(signal.SIGWINCH, partial(term_resize, term, grids))
+    def on_resize(signal, frame):
+        global do_refresh
+        do_refresh = True
+
+    signal.signal(signal.SIGWINCH, on_resize)

     score = 0
     with term.fullscreen(), term.cbreak(), term.hidden_cursor():
-        term_resize(term, grids)
-
         game_over = False
         while True:
-            draw_score(score, term, end=game_over)
+            if do_refresh:
+                draw_score(score, term, end=game_over)
+                term_resize(term, grids)
+                do_refresh = False

-            key = term.inkey()
+            key = term.inkey(0.5)
             if key == 'q' or game_over:
                 break
```

**Comment by rolfmorel** [comment]
Thank you for bug report.

I tried the same thing in xterm and urxvt on ArchLinux and I can not resize fast enough to reproduce the error, but having the signal handler run as little code as possible is a good idea and not calling any functions in the signal handler should take care of the re-entrance issue.

And by using a closure I will be able to solve two minor issues:
1. The score does not redraw after a resize, currently it only redraws after inkey() returns.
2. The grid will happily draw even when `term_resize()` just displayed a message saying that terminal is to small to accommodate the grid(s).

**Comment by jquast** [comment]
> I tried the same thing in xterm and urxvt on ArchLinux and I

Here's an idea:
  while [ 0 ]; do kill -WINCH <pid_of_macht> || break ; done

**Comment by rolfmorel** [comment]
I see the timeout for `inkey()` is needed to check every 0.5s if `do_refresh` is set.

A nicer solution would be if `inkey()` would propagate the `InterruptedError` exception and I would be able to catch it in user code and check immediately if `do_refresh` is set.
This would mean that the timeout wouldn't be necessary and that `term_resize()` is run immediately after the signal handler, but seeing as having `inkey()` continue after a signal handler has run is a sane default maybe it would be a idea to allow user code to specify if it would like to handle exceptions. This would probably mean adding a keyword argument, like `suppress=True` to `inkey()` and `kbhit()`.

**Comment by rolfmorel** [comment]
Just tried `while [ 0 ]; do pkill -WINCH -f macht || break ; done` for both xterm and urxvt and I can't get it to break.

**Comment by jquast** [comment]
it only reproduces for python3.

**Comment by rolfmorel** [comment]
I can't get it to reproduce on Linux with X11, but I was able to reproduce it with the virtual terminal (ctrl-alt-F1) when using python3.

I pushed a [improve_resize](https://github.com/polyphemus/macht/tree/improve_resize) branch which is based on your fix and uses the timeout. With the new commit macht no longer raises an exception when bombarded by SIGWINCH interrupts. The new commit also solves the two minor issues.

The only slight imperfection is a minor delay when resizing and the aesthetic issue of return every 0.5 seconds from a syscall.
